cmd/pebble/cmd_exec.go

```diff
@@ -0,0 +1,250 @@
+
+	"github.com/canonical/pebble/client"
+	"github.com/canonical/pebble/internal/logger"
+	"github.com/canonical/pebble/internal/ptyutil"
+)
+
+type cmdExec struct {
+	clientMixin
+	WorkingDir string        `long:"cwd"`
+	Env        []string      `long:"env"`
+	User       string        `long:"user"`
+	Group      string        `long:"group"`
+	Timeout    time.Duration `long:"timeout"`
+	Terminal   bool          `short:"t" long:"terminal"`
+	NoTerminal bool          `short:"T" long:"no-terminal"`
+	Positional struct {
+		Command string `positional-arg-name:"<command>" required:"1"`
+	} `positional-args:"yes"`
+}
+
+var execDescs = map[string]string{
+	"cwd":         "Working directory to run command in",
+	"env":         "Environment variable to set (in 'FOO=bar' format)",
+	"user":        "User name or ID to run command as",
+	"group":       "Group name or ID to run command as",
+	"timeout":     "Timeout after which to terminate command",
+	"terminal":    "Force pseudo-terminal allocation",
+	"no-terminal": "Disable pseudo-terminal allocation",
+}
+
+var shortExecHelp = "Execute a command and wait for it to finish"
+var longExecHelp = `
+The exec command executes a command via the Pebble API and waits for it to
+finish. Stdin is forwarded, and stdout and stderr are received. By default,
+exec's terminal mode is used if the terminal is a TTY (use -t/--terminal or
+-T/--no-terminal to override).
+
+To avoid confusion, exec options may be separated from the command and its
+arguments using "--", for example:
+
+pebble exec --timeout 10s -- echo -n foo bar
+`
+
+func (cmd *cmdExec) Execute(args []string) error {
+	if cmd.Terminal && cmd.NoTerminal {
+		return errors.New("can't pass -t and -T at the same time")
+	}
+
+	command := append([]string{cmd.Positional.Command}, args...)
+	logger.Debugf("Executing command %q", command)
+
+	// Set up environment variables.
+	env := make(map[string]string)
+	term, ok := os.LookupEnv("TERM")
+	if ok {
+		env["TERM"] = term
+	}
+	for _, kv := range cmd.Env {
+		parts := strings.SplitN(kv, "=", 2)
+		key := parts[0]
+		value := ""
+		if len(parts) == 2 {
+			value = parts[1]
+		}
+		env[key] = value
+	}
+
+	// Send UID if it looks like an integer, otherwise send username.
+	var user string
+	var userID *int
+	uid, err := strconv.Atoi(cmd.User)
+	if err != nil {
+		user = cmd.User
+	} else {
+		userID = &uid
+	}
+
+	// Send GID if it looks like an integer, otherwise send group name.
+	var group string
+	var groupID *int
+	gid, err := strconv.Atoi(cmd.Group)
+	if err != nil {
+		group = cmd.Group
+	} else {
+		groupID = &gid
+	}
+
+	// Specify UseTerminal if -t/--terminal is given, or if both stdin and
+	// stdout are TTYs.
+	stdinIsTerminal := ptyutil.IsTerminal(unix.Stdin)
+	stdoutIsTerminal := ptyutil.IsTerminal(unix.Stdout)
+	var useTerminal bool
+	if cmd.Terminal {
+		useTerminal = true
+	} else if cmd.NoTerminal {
+		useTerminal = false
+	} else {
+		useTerminal = stdinIsTerminal && stdoutIsTerminal
```

Comment:
See also below. I think these semantics are okay, and simple enough to explain: if stdin or stdout are being redirected, don't use a terminal (unless `-t` is given). For what it's worth, the default logic `ssh` uses is just `useTerminal = stdinIsTerminal` (see [here](https://github.com/openssh/openssh-portable/blob/9b2ee74e3aa8c461eb5552a6ebf260449bb06f7e/ssh.c#L1347-L1348)).

In any case, I think it's extremely rare you'd want to pipe stdin to (say) `nano` (on my system at least `echo foo | nano` gives an error). So it seems to me the current behavior (also the lxc behaviour) is reasonable.

Reply:
This still looks awkward for the reason explained above. Terminals are not just for interactive applications in the sense of an editor. Many of the commands we regularly run actually have a distinct behavior once a terminal is available. Try it out: run "ls", and then run "ls | cat", and then run "echo | ls cat".

So there's a clear a real motivation for us to preserve a terminal if we can output to a terminal. Is there a benefit in not having the terminal based just on stdin?

Reply:
Do you mean just based on stdout? Because it should use a terminal when running `ls` or `echo | ls`, but not `ls | cat`, right?

In any case, I tried to adjust the code to do `useTerminal = stdoutIsTerminal` and it seems to work, but then `echo | pebble exec -- ls` exits with error code 129 due to a `POLLNVAL` error deep in the wsutil PTY proxying code. I'll leave this as an open issue, but for now I'm going to leave it as is and we can concentrate on the rest of the PR.

Reply:
> Because it should use a terminal when running ls or echo | ls, but not ls | cat, right?

That's right. :+1:

> but then echo | pebble exec -- ls exits with error code 129 due to a POLLNVAL
(...)
> I'll leave this as an open issue

Thanks, we definitely want that working.

Reply:
I realize this is a bug, but I would like to get this PR in and fix in a separate PR. LXD has this issue as well, and it's very unlikely to happen in practice (it obviously doesn't come up for `lxc exec`), so it doesn't seem urgent, and would help us get this PR through to make progress. If you don't disagree, I'll open a follow-up issue.

Reply:
I've created PR https://github.com/canonical/pebble/pull/78 to revert this behavior, with explanation there.